File: src/main/java/no/nav/veilarbaktivitet/aktivitetskort/util/ZonedOrNorwegianDateTimeDeserializer.java
package no.nav.veilarbaktivitet.aktivitetskort.util;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.datatype.jsr310.deser.LocalDateTimeDeserializer;
import com.fasterxml.jackson.datatype.jsr310.deser.key.ZonedDateTimeKeyDeserializer;

import java.io.IOException;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;


public class ZonedOrNorwegianDateTimeDeserializer extends JsonDeserializer<ZonedDateTime> {
    private static final ZonedDateTimeKeyDeserializer zonedDateTimeKeyDeserializer = ZonedDateTimeKeyDeserializer.INSTANCE;
    private static final LocalDateTimeDeserializer localDateTimeDeserializer = new LocalDateTimeDeserializer(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
    @Override
    public ZonedDateTime deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
        String dateText = jsonParser.getText();
        String localDateTimeRegex = "\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d{1,6})?$";

        if (dateText.matches(localDateTimeRegex)) {
            return localDateTimeDeserializer.deserialize(jsonParser, deserializationContext).atZone(ZoneId.of("Europe/Oslo"));
        } else {
            return (ZonedDateTime) zonedDateTimeKeyDeserializer.deserializeKey(dateText, deserializationContext);
        }
    }
}

File: src/main/kotlin/no/nav/veilarbaktivitet/aktivitetskort/util/ZonedOrNorwegianDateTimeDeserializer.kt
package no.nav.veilarbaktivitet.aktivitetskort.util

import tools.jackson.core.JsonParser
import tools.jackson.databind.DeserializationContext
import java.io.IOException
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import tools.jackson.databind.ValueDeserializer
import tools.jackson.databind.ext.javatime.deser.LocalDateTimeDeserializer
import tools.jackson.databind.ext.javatime.deser.key.ZonedDateTimeKeyDeserializer

class ZonedOrNorwegianDateTimeDeserializer : ValueDeserializer<ZonedDateTime>() {
    @Throws(IOException::class)
    override fun deserialize(jsonParser: JsonParser, deserializationContext: DeserializationContext): ZonedDateTime {
        val dateText = jsonParser.string
        val localDateTimeRegex = "\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d{1,6})?$"
        return if (dateText.matches(localDateTimeRegex.toRegex())) {
            localDateTimeDeserializer.deserialize(
                jsonParser,
                deserializationContext
            ).atZone(ZoneId.of("Europe/Oslo"))
        } else {
            zonedDateTimeKeyDeserializer.deserializeKey(
                dateText,
                deserializationContext
            ) as ZonedDateTime
        }
    }

    companion object {
        private val zonedDateTimeKeyDeserializer = ZonedDateTimeKeyDeserializer.INSTANCE
        private val localDateTimeDeserializer = LocalDateTimeDeserializer(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
    }
}
